x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -874,4 +886,127 @@ public enum AuthenticationType {
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            if (false == super.equals(o)) return false;
+            RunAsUser runAsUser = (RunAsUser) o;
+            return authenticatingUser.equals(runAsUser.authenticatingUser);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(super.hashCode(), authenticatingUser);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append("RunAsUser[username=").append(principal());
+            sb.append(",roles=[").append(Strings.arrayToCommaDelimitedString(roles())).append("]");
+            sb.append(",fullName=").append(fullName());
+            sb.append(",email=").append(email());
+            sb.append(",metadata=");
+            sb.append(metadata());
+            if (enabled() == false) {
+                sb.append(",(disabled)");
+            }
+            sb.append(",authenticatingUser=[").append(authenticatingUser.toString()).append("]");
+            sb.append("]");
+            return sb.toString();
+        }
+    }
+
+    public static class AuthenticationSerializationHelper {
+
+        private AuthenticationSerializationHelper() {}
+
+        public static User readUserFrom(StreamInput input) throws IOException {
+            final boolean isInternalUser = input.readBoolean();
+            final String username = input.readString();
+            if (isInternalUser) {
+                if (SystemUser.is(username)) {
+                    return SystemUser.INSTANCE;
+                } else if (XPackUser.is(username)) {
+                    return XPackUser.INSTANCE;
+                } else if (XPackSecurityUser.is(username)) {
+                    return XPackSecurityUser.INSTANCE;
+                } else if (SecurityProfileUser.is(username)) {
+                    return SecurityProfileUser.INSTANCE;
+                } else if (AsyncSearchUser.is(username)) {
+                    return AsyncSearchUser.INSTANCE;
+                }
+                throw new IllegalStateException("user [" + username + "] is not an internal user");
+            }
+            return partialReadUserFrom(username, input);
+        }
+
+        public static void writeUserTo(User user, StreamOutput output) throws IOException {
+            if (SystemUser.is(user)) {
+                output.writeBoolean(true);
+                output.writeString(SystemUser.NAME);
+            } else if (XPackUser.is(user)) {
+                output.writeBoolean(true);
+                output.writeString(XPackUser.NAME);
+            } else if (XPackSecurityUser.is(user)) {
+                output.writeBoolean(true);
+                output.writeString(XPackSecurityUser.NAME);
+            } else if (SecurityProfileUser.is(user)) {
```

Comment:
Do we need BWC for this given the feature is still beta? I had assumed that we don't need it. For example, the profile index mappings changed in both 8.2 and 8.3 in a non-compatible way. It is unlikely things would work in a mixed cluster even without this new internal user. That said, maybe we should have phased the changes better so that they are BWC all the way. Unfortunately it was not the case. What do you think?

Reply:
But what happens when someone does a rolling upgrade from 7.17 to 8.7 ? Will the new internal user get passed around in cross-node requests and cause failures on the 7.17 nodes?

Reply:
If they call User Profile APIs on the 8.7 node, then yes, it is possible that they can see the failure on a 7.17 node. A BWC for this internal user will not fix the failure entirely. But one may argue that the error message could be slightly more friendy?

Instead of a mysterious serialisation failure (`user [_security_profile] is not an internal user`), it could be a 404 index_not_found (`no such index [_security]`) or 400 (`request body is required`) or some other errors. Now I actually don't know which one is less friendly. The serialisation failure is probably more mysterious but it fails consistently and mentions `profile` in the failure message, while the other option could be more unpredictable.

Reply:
Sorry I have to correct myself: the 400 or 404 error is what user can get on the REST layer. Since this talks about transport layer, I think the error should always be something along the line of action handler not found, which is arguably a more friend error message. I think this is enough to convince me to handle BWC for this user. I will raise a separate PR. Thanks for the discussion.

Reply:
I raised https://github.com/elastic/elasticsearch/pull/86345
Btw, the BWC with the above PR is better than I expected. As long as the request hits nodes with newer version first, it should just work because requests that actually got sent across nodes are basic index and search. Feature specific actions are all handled locally.